File: test/scripts/memory-allocation.js
#!/usr/bin/env node
/**
 * ANTI-PATTERN: Excessive allocations (no pooling)
 * Resume: "Use Memory Pooling" - Many allocations cause GC pressure.
 *
 * Run: node memory-allocation.js
 * Measure: node --expose-gc memory-allocation.js, or --trace-gc
 */

const N = 500_000;

console.log('ANTI-PATTERN: Excessive allocations (no pooling)');
console.log(`Allocating ${N} small objects`);

// BAD: New object per iteration
let start = process.hrtime.bigint();
const items = [];
for (let i = 0; i < N; i++) {
  items.push({ id: i, value: String(i).repeat(10) });
}
let elapsed = Number(process.hrtime.bigint() - start) / 1e6;

const total = items.reduce((s, x) => s + x.id, 0);
console.log(`No pooling: ${elapsed.toFixed(0)} ms, ${N} allocations, sum=${total}`);
console.log('Fix: use object pooling, reuse buffers, avoid allocations in hot path');

File: test/scripts/memory-allocation.py
#!/usr/bin/env python3
"""
ANTI-PATTERN: Excessive allocations (no pooling)
Resume: "Use Memory Pooling" - Many allocations cause GC pressure,
fragmentation, and overhead.

Run: python memory-allocation.py
Measure: memory_profiler, tracemalloc
"""

import time

N = 500_000


def main():
    print("ANTI-PATTERN: Excessive allocations (no pooling)")
    print(f"Allocating {N} small objects")

    # BAD: New list/dict per iteration
    start = time.perf_counter()
    items = []
    for i in range(N):
        items.append({"id": i, "value": str(i) * 10})
    elapsed = time.perf_counter() - start

    total = sum(x["id"] for x in items)
    print(f"No pooling: {elapsed*1000:.0f} ms, {N} allocations, sum={total}")

    # With reuse (fewer allocations) - just for comparison
    start = time.perf_counter()
    items2 = []
    buf = {}
    for i in range(N):
        buf["id"] = i
        buf["value"] = str(i) * 10
        items2.append(buf.copy())  # Still allocates copy
    elapsed2 = time.perf_counter() - start
    print(f"With copy:  {elapsed2*1000:.0f} ms")
    print("Fix: use ObjectPool, ArrayPool, or reuse buffers")


if __name__ == "__main__":
    main()
